### ISSUE TITLE
`torch.{addcmul,addcdiv}` triggers INTERNAL ASSERT FAILED

### ISSUE THREAD

**Issue by ilovepytorch** [pytorch/pytorch#70486]
### 🐛 Describe the bug

It seems that when the `tensor2` has complex dtype, it will trigger internal assert fail.

```python
import torch
arg_1 = torch.rand([5, 5], dtype=torch.float64)
arg_2 = torch.rand([5, 5], dtype=torch.float64)
arg_3 = torch.rand([1, 5], dtype=torch.complex128)
res = torch.addcmul(arg_1,arg_2,arg_3)
# RuntimeError: !(has_different_input_dtypes && !config.promote_inputs_to_common_dtype_ && (has_undefined_outputs || config.enforce_safe_casting_to_output_ || config.cast_common_dtype_to_outputs_))INTERNAL ASSERT FAILED at "../aten/src/ATen/TensorIterator.cpp":331, please report a bug to PyTorch. 
```

### Versions

pytorch: 1.10.1

cc @ezyang @gchanan @zou3519 @bdhirsh @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved @nairbv @albanD @soulitzer

**Comment by ilovepytorch** [comment]
`torch.addcdiv` also has this problem

```python
import torch
arg_1 = torch.rand([1], dtype=torch.complex128)
arg_2 = torch.rand([5, 5, 1], dtype=torch.complex128)
arg_3 = torch.rand([1, 3], dtype=torch.float32)
res = torch.addcdiv(arg_1,arg_2,arg_3,)
# RuntimeError: !(has_different_input_dtypes && !config.promote_inputs_to_common_dtype_ && (has_undefined_outputs || config.enforce_safe_casting_to_output_ || config.cast_common_dtype_to_outputs_))INTERNAL ASSERT FAILED at "../aten/src/ATen/TensorIterator.cpp":331, please report a bug to PyTorch. 
```

**Comment by albanD** [comment]
Hi!
Thanks for the report.
Just to let you know, there is no need to "cc" people on the issue. The bot will cc the right people automatically based on the labels that are added.

**Comment by Lezcano** [comment]
cc @nkaretnikov you might want to have a stab at this one once you finish the PR you're working on :)

**Comment by anjali411** [comment]
I think the issue is just that all the input tensors are not of the same dtype, no? Seems consistent with our policy for other ops

**Comment by mruberry** [comment]
> I think the issue is just that all the input tensors are not of the same dtype, no? Seems consistent with our policy for other ops. e.g., this fails too:
> 
> ```python
> arg_1 = torch.rand([1], dtype=torch.float64)
> arg_2 = torch.rand([5, 5, 1], dtype=torch.float64)
> arg_3 = torch.rand([1, 3], dtype=torch.float32)
> res = torch.addcdiv(arg_1,arg_2,arg_3,)
> ```

It shouldn't trigger an internal assert, though, but produce an error message that type promotion isn't supported. Also we could support type promotion for addcmul and addcdiv since they're elementwise operators.

**Comment by anjali411** [comment]
> It shouldn't trigger an internal assert, though, but produce an error message that type promotion isn't supported.

oh gotcha makes sense

**Comment by ilovepytorch** [comment]
@mruberry @albanD Hi, I find that there are many APIs that will trigger the INTERNAL ASSERT FAILER when the inputs have different types.

```python
torch.ge
torch.le
torch.gt
torch.lt
torch.ne
torch.eq
torch.igamma
torch.igammac
torch.xlogy
torch.logical_and
torch.logical_or
torch.logical_xor
torch.bucketize
torch.searchsorted
torch.dist
torch.ldexp
torch.sub
torch.add
torch.mul
torch.div
```

A case to trigger:
```python
import torch
results = dict()
input_tensor = torch.rand([1], dtype=torch.float64)
input = input_tensor.clone()
other_tensor = torch.rand([1], dtype=torch.complex32)
other = other_tensor.clone()
try:
  results["res_1"] = torch.add(input, other, )
except Exception as e:
  results["err_1"] = "ERROR:"+str(e)
try:
  results["res_2"] = torch.sub(input_tensor.clone(),other_tensor.clone(),)
except Exception as e:
  results["err_2"] = "ERROR:"+str(e)

print(results)
# {'err_1': 'ERROR:common_dtype_ != ScalarType::UndefinedINTERNAL ASSERT FAILED at "../aten/src/ATen/TensorIterator.cpp":236, please report a bug to PyTorch. ', 'err_2': 'ERROR:common_dtype_ != ScalarType::UndefinedINTERNAL ASSERT FAILED at "../aten/src/ATen/TensorIterator.cpp":236, please report a bug to PyTorch. '}
```

I am still finding more such APIs. Hope this can help you!

**Comment by nkaretnikov** [comment]
i'm putting this in the pipeline of issues i'll be working on.

**Comment by nkaretnikov** [comment]
started working on this

**Comment by nkaretnikov** [comment]
so i remembered that i had added this https://github.com/pytorch/pytorch/pull/72324.

while testing on `gh/nkaretnikov/28/orig` (that pr), both addcmul and addcdiv don't crash.

it works because that patch adds more config options to various *build helpers used by TI, so the assert is not triggered.

could you take a look at that pr and tell me whether it's something that makes sense to have?

there's also an issue about that here: https://github.com/pytorch/pytorch/issues/70238

(i haven't tested yet whether the behavior of operators is somehow affected by that patch, but it only has one ci test failure)

**Comment by nkaretnikov** [comment]
if that pr is okay, i'll clean it up and make sure it passes ci tests and replace this internal assert with TORCH_CHECK + a custom error message. how does it sound?

**Comment by pearu** [comment]
"Cannot" sounds like it is "impossible" while, IIUC, the issue here is that the feature has not been implemented.
So, how about: "Promoting types of the given arguments not yet supported."?

**Comment by mruberry** [comment]
I don't think we need to change TI to fix the cases mentioned. It would be great to fix them all, but they should be fixable with simple updates to verify all tensor dtypes are the same or to dispatch on `common_dtype`. 

I can't figure out to trigger the internal asserts on most of these. How did you trigger the internal assert for `torch.add`? Does it require using complex32? If so that's not very surprising and we shipped with it disabled in 1.11 for this reason. cc @kshitij12345. 

We have spot checks for type promotion for elementwise binary operators: 

https://github.com/pytorch/pytorch/blob/99bcadced431cbe040653f6217a827ba97b11065/test/test_binary_ufuncs.py#L566

If those spot checks are insufficient we should be sure to update them. And we should definitely update them as part of improving our complex32 support. 

It would be nice to add a new test class, test_elementwise_ternary.py, for ops like addcmul and addcdiv. A generic test for type promotion could be written for them. Do we have any elementwise quaternary ops?

**Comment by nkaretnikov** [comment]
@mruberry okay, i'll take a look at individual ops. we can look at the ti thing separately later

**Comment by nkaretnikov** [comment]
@mruberry forgot to respond to this bit:

> Do we have any elementwise quaternary ops?

well, at least this one (margin_ranking_loss):

https://github.com/pytorch/pytorch/pull/72321/files

**Comment by nkaretnikov** [comment]
@mruberry here's my test for `add` with all types (?). indeed, it seems to be that only `complex32` fails:

```
% python ../test_diff_dtype_assert_all.py   | grep add
test 3: add failed, types: torch.float16, torch.complex32
test 10: add failed, types: torch.float32, torch.complex32
test 17: add failed, types: torch.float64, torch.complex32
test 21: add failed, types: torch.complex32, torch.float16
test 22: add failed, types: torch.complex32, torch.float32
test 23: add failed, types: torch.complex32, torch.float64
test 24: add failed, types: torch.complex32, torch.complex32
test 25: add failed, types: torch.complex32, torch.complex64
test 26: add failed, types: torch.complex32, torch.complex128
test 27: add failed, types: torch.complex32, torch.bfloat16
test 31: add failed, types: torch.complex64, torch.complex32
test 38: add failed, types: torch.complex128, torch.complex32
test 45: add failed, types: torch.bfloat16, torch.complex32
```

```py
import torch

# see AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_AND_QINTS
types = (
    torch.uint8,       # 0
    torch.int8,        # 1
    torch.int16,       # 2
    torch.int,         # 3
    torch.int64,       # 4
    torch.half,        # 5
    torch.float,       # 6
    torch.double,      # 7
    torch.complex32,   # 8
    torch.complex64,   # 9
    torch.complex128,  # 10
    torch.bool,        # 11
    torch.qint8,       # 12
    torch.quint8,      # 13
    torch.qint32,      # 14
    torch.bfloat16,    # 15
    torch.quint4x2,    # 16
    torch.quint2x4,    # 17
)

i = 0
for t1 in types:
    for t2 in types:
        try:
            x = torch.rand([1], dtype=t1)
        except Exception as e:
            print(f"test {i}: rand failed for x, type: {t1}")
            continue

        try:
            y = torch.rand([1], dtype=t2)
        except Exception as e:
            print(f"test {i}: rand failed for y, type: {t2}")
            continue

        try:
            res = torch.add(x, y)
            print(f"test {i}:", res)
        except Exception as e:
            print(f"test {i}: add failed, types: {t1}, {t2}")

        i += 1

```

**Comment by nkaretnikov** [comment]
@mruberry 

>  It would be great to fix them all, but they should be fixable with simple updates to verify all tensor dtypes are the same or to dispatch on common_dtype.

so... i've started looking at this. my current understanding (could be wrong) is that dispatching/promotion should be handled by TI implicitly, without extra effort. see the following discussion for reference:

https://github.com/pytorch/pytorch/pull/72321#discussion_r802073157

there was also an issue with structured and TI:

https://github.com/pytorch/pytorch/pull/72321#discussion_r801127733

i noticed addcmul is using both TI and structured and was ported fairly recently, so i decided to (manually) revert the said patch and see if the behavior would be different. i put the code here:

https://github.com/nkaretnikov/pytorch/commit/f0d3f76b0fdb17c90849824d331723667e71c2f6

the TI part here should be similar, but this reverts the structured part.

with the original addcmul test posted above:

```py
import torch
arg_1 = torch.rand([5, 5], dtype=torch.float64)
arg_2 = torch.rand([5, 5], dtype=torch.float64)
arg_3 = torch.rand([1, 5], dtype=torch.complex128)
res = torch.addcmul(arg_1,arg_2,arg_3)
# RuntimeError: !(has_different_input_dtypes && !config.promote_inputs_to_common_dtype_ && (has_undefined_outputs || config.enforce_safe_casting_to_output_ || config.cast_common_dtype_to_outputs_))INTERNAL ASSERT FAILED at "../aten/src/ATen/TensorIterator.cpp":331, please report a bug to PyTorch. 
```

i get this now:

```
RuntimeError: Found dtype ComplexDouble but expected Double
```

which seems to correspond to this, happens after the internal_assert reported earlier:

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorIterator.cpp#L411

```cpp
  // Checks that either the computation type is computable or unneeded
  TORCH_INTERNAL_ASSERT(!(has_different_input_dtypes && !config.promote_inputs_to_common_dtype_ &&
                        (has_undefined_outputs || config.enforce_safe_casting_to_output_ ||
                        config.cast_common_dtype_to_outputs_)));

  // Checks that all inputs and defined outputs are the same dtype, if requested
  if (config.check_all_same_dtype_ &&
      (has_different_input_dtypes || has_different_output_dtypes ||
      (common_dtype_ != output_dtype && output_dtype != ScalarType::Undefined))) {
    // Throws an informative error message
    for (auto& op : operands_) {
      if (!op.tensor_base().defined()) {
        continue;
      }

      TORCH_CHECK(op.target_dtype == common_dtype_,
                  "Found dtype ", op.target_dtype, " but expected ", common_dtype_);
    }
  }
```

so the behavior does change.

now, i'm not exactly sure how to handle this given that:
- implicit type promotion by TI doesn't seem to work properly
- manual promotion is an anti-pattern
- structured is necessary for meta kernels.

do we just use something like this despite this being an anti-pattern?

https://github.com/pytorch/pytorch/pull/72321#discussion_r801008311

apologies if this doesn't make sense, i'm trying to understand promotion rules/ti/structured based on various random pieces of information gathered while working on similar things

**Comment by mruberry** [comment]
It's because addcmul is dispatching on iter.dtype() and not iter.common_dtype():

https://github.com/pytorch/pytorch/blob/213ac4e59cf5ce4c53780cec3f96dbb82aae1f4d/aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp#L36

`iter.dtype()` is (confusingly) the datatype of the output, `iter.common_dtype()` is the "computation type" determined by type promotion.

**Comment by mruberry** [comment]
For add I wouldn't worry about complex32 failures, @kshitij12345 is working on complex32 support now

**Comment by nkaretnikov** [comment]
for reference, i actually asked about dtype vs common_dtype a while back:

https://github.com/pytorch/pytorch/issues/46302#issuecomment-1000090889

**Comment by nkaretnikov** [comment]
@mruberry so i've read the above, but i'm still not sure what to do here, sorry.

e.g., with both dtype(0) and common_dtype(), i get this (i changed the cpu kernel code):

```cpp
  ScalarType dtype = iter.dtype(0);
  if (iter.dtype() == kBFloat16) {
  // ScalarType dtype = iter.common_dtype();
  // if (dtype == kBFloat16) {
```

(this is also used in the else case in the function)

```cpp
/*
has_different_input_dtypes 1
config.promote_inputs_to_common_dtype_ 0
has_undefined_outputs 1
config.enforce_safe_casting_to_output_ 0
config.cast_common_dtype_to_outputs_ 0
*/

  std::cout << "has_different_input_dtypes " << has_different_input_dtypes << std::endl;
  std::cout << "config.promote_inputs_to_common_dtype_ " << config.promote_inputs_to_common_dtype_ << std::endl;
  std::cout << "has_undefined_outputs " << has_undefined_outputs << std::endl;
  std::cout << "config.enforce_safe_casting_to_output_ " << config.enforce_safe_casting_to_output_ << std::endl;
  std::cout << "config.cast_common_dtype_to_outputs_ " << config.cast_common_dtype_to_outputs_ << std::endl;

  // Checks that either the computation type is computable or unneeded
  TORCH_INTERNAL_ASSERT(
    !(has_different_input_dtypes &&
      !config.promote_inputs_to_common_dtype_ &&
    (has_undefined_outputs ||
     config.enforce_safe_casting_to_output_ ||
     config.cast_common_dtype_to_outputs_)));
```

so i need to alter these values somehow to make it past this check. one way to do it, as i mentioned before, is to set the defaults in TI. but since you said that's not an option, i'm not really sure how to do it. i've looked at some other kernels as well, i don't really see any differences. e.g., compared to smooth_l1_loss.

any examples of kernels that do it right? or just general guidance on how to do it?

**Comment by mruberry** [comment]
Because 

https://github.com/pytorch/pytorch/blob/16554bec1b3b5284a9786be6a87dc0ddc3cb2e7c/aten/src/ATen/TensorIterator.cpp#L937

doesn't allow opting-in to type promotion like 

https://github.com/pytorch/pytorch/blob/16554bec1b3b5284a9786be6a87dc0ddc3cb2e7c/aten/src/ATen/TensorIterator.cpp#L957

You need to extend the TensorIterator sugar for ternary operators, ask for type promotion support, then dispatch on `common_dtype()`.

**Comment by nkaretnikov** [comment]
@mruberry wait, i'm confused now. i mentioned a patch before that does it (with some api refactoring) here:

https://github.com/pytorch/pytorch/issues/70486#issuecomment-1047221420

is it not the same thing?

in that patch, BINARY_OP_CONFIG is renamed to ELEMENTWISE_OP_CONFIG since it's a better name for 3+ arg ops. otherwise, it looks exactly like what you're proposing, no?

lmk if i missed anything and apologies for not making it more clear in my earlier msg

**Comment by Lezcano** [comment]
I agree with @nkaretnikov. I believe that his PR https://github.com/pytorch/pytorch/pull/72324 solves this issue.

**Comment by mruberry** [comment]
Sure but there's a much simpler and more targeted fix that can be made here. #72324 contains a lot of additional work.

**Comment by nkaretnikov** [comment]
okay, i'll try the targeted fix

**Comment by nkaretnikov** [comment]
created a pr with proposed changes: https://github.com/pytorch/pytorch/pull/74234

note: just using `common_dtype()` is not enough to fix this, so i had to add the options from `BINARY_OP_CONFIG` when initializing TI config. i didn't do it for all ternary operators because it's supposed to be a local fix

**Comment by Jup11ter** [comment]
There is a similar crash `INTERNAL ASSERT FAILED` raised on `torch.Tensor.{addcmul,addcdiv}` in 1.11.0 and 1.12.0, which raises:
```
Traceback (most recent call last):
  File "test.py", line 14, in <module>
    res = test()
  File "test.py", line 12, in test
    tmp_result= func_cls(arg_1,arg_2,arg_3)
RuntimeError: !(has_different_input_dtypes && !config.promote_inputs_to_common_dtype_ && (has_undefined_outputs || config.enforce_safe_casting_to_output_ || config.cast_common_dtype_to_outputs_)) INTERNAL ASSERT FAILED at "/opt/conda/conda-bld/pytorch_1656352464346/work/aten/src/ATen/TensorIterator.cpp":405, please report a bug to PyTorch. 
```

This bug is fixed in 1.13.0. Considering the relation between `torch.Tensor.addcmul` and `torch.addcmul`, the bug on `torch.Tensor.{addcmul,addcdiv}` should also be repaired by one of these [commits](https://github.com/pytorch/pytorch/commit/a04ffb7f896f34b7afc365f15e846cf67b8b55d7) in this issue.



```python
import torch

# func_cls=torch.Tensor.addcmul
func_cls=torch.Tensor.addcdiv

arg_1 = torch.rand([5, 5], dtype=torch.float64)
arg_2 = torch.rand([5, 5], dtype=torch.float64)
arg_3 = torch.rand([1, 5], dtype=torch.complex128)

def test():
	tmp_result= func_cls(arg_1,arg_2,arg_3)
	return tmp_result
res = test()

```
